log-manager/src/main/java/io/airlift/log/JsonFormatter.java

```diff
@@ -3,20 +3,27 @@
 import com.fasterxml.jackson.core.JsonFactory;
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
 import io.airlift.json.JsonCodec;
 import io.airlift.json.JsonCodecFactory;
 import io.airlift.json.ObjectMapperProvider;
 
 import java.io.IOException;
 import java.io.StringWriter;
-import java.util.logging.Formatter;
+import java.util.Map;
 import java.util.logging.LogRecord;
 
 public class JsonFormatter
-        extends Formatter
+        extends java.util.logging.Formatter
 {
     private static final JsonCodec<JsonRecord> CODEC = new JsonCodecFactory(new ObjectMapperProvider()).jsonCodec(JsonRecord.class);
     private static final JsonFactory jsonFactory = new JsonFactory();
+    private final Map<String, String> logAnnotations;
+
+    public JsonFormatter(Map<String, String> logAnnotations)
+    {
+        this.logAnnotations = logAnnotations == null ? null : ImmutableMap.copyOf(logAnnotations);
```

Comment:
Instead of using nulls, why not just use empty maps everywhere. It will make it easier and less error prone.

Reply:
I'd need to special case the JsonRecord getter to return null then or else all the json log lines would contain an empty map.

Reply:
Yeah, but it's better to special-case for JsonRecord when it gets constructed than special case everywhere